File: backend/app/models/backtest.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Backtest(Base):
    __tablename__ = "backtests"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    stock_id: Mapped[int | None] = mapped_column(ForeignKey("stocks.id"), nullable=True)
    sector_id: Mapped[int | None] = mapped_column(ForeignKey("sectors.id"), nullable=True)
    mode: Mapped[str] = mapped_column(String(20), nullable=False)  # technical, full
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")  # pending, running, completed, failed
    start_date: Mapped[datetime] = mapped_column(DateTime(timezone=False), nullable=False)
    end_date: Mapped[datetime] = mapped_column(DateTime(timezone=False), nullable=False)
    starting_capital: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False, default=10000)
    min_signal_strength: Mapped[str] = mapped_column(String(20), nullable=False, default="moderate")

    # Result metrics (populated on completion)
    total_return_pct: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    annualized_return_pct: Mapped[float | None] = mapped_column(Numeric(10, 4), nullable=True)
    sharpe_ratio: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)
    max_drawdown_pct: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)
    win_rate_pct: Mapped[float | None] = mapped_column(Numeric(6, 4), nullable=True)
    total_trades: Mapped[int | None] = mapped_column(Integer, nullable=True)
    avg_win_pct: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)
    avg_loss_pct: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)
    best_trade_pct: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)
    worst_trade_pct: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)
    final_equity: Mapped[float | None] = mapped_column(Numeric(14, 2), nullable=True)

    # Equity curve as JSON (written once on completion)
    equity_curve: Mapped[str | None] = mapped_column(Text, nullable=True)

    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    celery_task_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Relationships
    user = relationship("User")
    stock = relationship("Stock")
    sector = relationship("Sector")
    trades = relationship("BacktestTrade", back_populates="backtest", cascade="all, delete-orphan")


class BacktestTrade(Base):
    __tablename__ = "backtest_trades"

    id: Mapped[int] = mapped_column(primary_key=True)
    backtest_id: Mapped[int] = mapped_column(
        ForeignKey("backtests.id", ondelete="CASCADE"), nullable=False, index=True
    )
    ticker: Mapped[str] = mapped_column(String(10), nullable=False)
    action: Mapped[str] = mapped_column(String(10), nullable=False)  # buy, sell
    trade_date: Mapped[datetime] = mapped_column(DateTime(timezone=False), nullable=False)
    price: Mapped[float] = mapped_column(Numeric(12, 4), nullable=False)
    shares: Mapped[float] = mapped_column(Numeric(14, 6), nullable=False)
    position_value: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    portfolio_equity: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    signal_score: Mapped[float] = mapped_column(Numeric(8, 5), nullable=False)
    signal_direction: Mapped[str] = mapped_column(String(20), nullable=False)
    signal_strength: Mapped[str] = mapped_column(String(20), nullable=False)
    return_pct: Mapped[float | None] = mapped_column(Numeric(8, 4), nullable=True)

    backtest = relationship("Backtest", back_populates="trades")

File: frontend/src/types/backtest.ts
export interface BacktestConfig {
  ticker?: string;
  sector_name?: string;
  start_date: string;
  end_date: string;
  starting_capital?: number;
  mode?: "technical" | "full";
  min_signal_strength?: "moderate" | "strong";
  commission_pct?: number;
  slippage_pct?: number;
  position_size_pct?: number;
  stop_loss_pct?: number | null;
  take_profit_pct?: number | null;
  benchmark_ticker?: string;
}

export interface BacktestSummary {
  id: number;
  ticker: string | null;
  sector_name: string | null;
  mode: string;
  status: "pending" | "running" | "completed" | "failed";
  start_date: string;
  end_date: string;
  starting_capital: number;
  min_signal_strength: string;
  commission_pct: number | null;
  slippage_pct: number | null;
  position_size_pct: number | null;
  stop_loss_pct: number | null;
  take_profit_pct: number | null;
  benchmark_ticker: string | null;
  total_return_pct: number | null;
  annualized_return_pct: number | null;
  sharpe_ratio: number | null;
  max_drawdown_pct: number | null;
  win_rate_pct: number | null;
  total_trades: number | null;
  avg_win_pct: number | null;
  avg_loss_pct: number | null;
  best_trade_pct: number | null;
  worst_trade_pct: number | null;
  final_equity: number | null;
  benchmark_total_return_pct: number | null;
  benchmark_annualized_return_pct: number | null;
  alpha: number | null;
  beta: number | null;
  error_message: string | null;
  created_at: string;
  completed_at: string | null;
}

export interface EquityPoint {
  date: string;
  equity: number;
}

export interface BacktestTrade {
  id: number;
  ticker: string;
  action: "buy" | "sell";
  trade_date: string;
  price: number;
  shares: number;
  position_value: number;
  portfolio_equity: number;
  signal_score: number;
  signal_direction: string;
  signal_strength: string;
  return_pct: number | null;
  exit_reason: string | null;
}

export interface BacktestDetail extends BacktestSummary {
  equity_curve: EquityPoint[];
  trades: BacktestTrade[];
  benchmark_equity_curve: EquityPoint[] | null;
}
